clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java

```diff
@@ -381,18 +377,23 @@ public boolean matches(ClientRequest request) {
             }
         }, new HeartbeatResponse(Errors.NONE.code()).toStruct(), coordinator);
 
+        // heartbeat interval is 2 seconds
+        time.sleep(heartbeatIntervalMs);
+        Thread.sleep(heartbeatIntervalMs);
```

Comment:
How about letting the background thread constructor gets a `Runnable` and moving the `run` method to the Runnable object? Then for testing we can just trigger the `Runnable.run` method with mock time?

BTW, could we also leverage on `ShutdownableThread` from connect for the shutdown logic of heartbeat thread?